pallets/xcm-asset-manager/src/lib.rs

```diff
@@ -0,0 +1,423 @@
+			AssetTypeId::<T>::insert(&asset_type, &asset_id);
+
+			Self::deposit_event(Event::<T>::ForeignAssetTypeRegistered { asset_id, asset_type });
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+
+			Ok(())
+		}
+
+		/// Relocate asset id.
+		/// Can only be larger than current assignment.
+		#[pallet::weight(T::WeightInfo::relocate_foreign_asset_id())]
+		pub fn relocate_foreign_asset_id(
+			origin: OriginFor<T>,
+			new_asset_tracker: T::AssetId,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+			let next_asset_id = new_asset_tracker
+				.checked_add(&One::one())
+				.ok_or(Error::<T>::AssetIdLimitReached)?;
+			let old_existing_asset_id = ForeignAssetTracker::<T>::get();
+			if next_asset_id > old_existing_asset_id {
+				Self::deposit_event(Event::<T>::ForeignAssetTrackerUpdated {
+					old_asset_tracker: old_existing_asset_id,
+					new_asset_tracker,
+				});
+				ForeignAssetTracker::<T>::put(new_asset_tracker);
+			}
+			Ok(())
+		}
+
+		// Update asset metadata
+		#[pallet::weight(T::WeightInfo::update_foreign_asset_metadata())]
+		pub fn update_foreign_asset_metadata(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<BalanceOf<T>>,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			ensure!(AssetIdType::<T>::contains_key(&asset_id), Error::<T>::AssetIdDoesNotExist);
+			AssetIdMetadata::<T>::insert(&asset_id, &metadata);
+
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+			Ok(())
+		}
+
+		/// Change the amount of units we are charging per execution second
+		/// for a given ForeignAssetType
+		/// 0 means not support
+		#[pallet::weight(T::WeightInfo::set_asset_units_per_second())]
+		pub fn set_asset_units_per_second(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			units_per_second: u128,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			ensure!(AssetIdType::<T>::contains_key(&asset_id), Error::<T>::AssetIdDoesNotExist);
+
+			AssetIdUnitsPerSecond::<T>::insert(&asset_id, &units_per_second);
+
+			Self::deposit_event(Event::UnitsPerSecondChanged { asset_id, units_per_second });
+			Ok(())
+		}
+
+		/// Add the xcm type mapping for a existing assetId, other assetType still exists if any.
+		#[pallet::weight(T::WeightInfo::add_asset_type())]
+		pub fn add_asset_type(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			new_asset_type: T::ForeignAssetType,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			ensure!(AssetIdType::<T>::contains_key(&asset_id), Error::<T>::AssetIdDoesNotExist);
+			ensure!(
+				!AssetTypeId::<T>::contains_key(&new_asset_type),
+				Error::<T>::AssetAlreadyExists
+			);
+
+			// Insert new asset type info
+			AssetIdType::<T>::insert(&asset_id, &new_asset_type);
+			AssetTypeId::<T>::insert(&new_asset_type, &asset_id);
```

Comment:
Here is a bit misleading IMHO, although the result should be as expected(the first insert updates the value, and the second insert() adds a new k/v pair). Use `mutate` would be much more understandable.

Reply:
<AssetIdType<T>>::try_mutate(id, |type| {
                type = new_type;
            }).map_err(|_| return Error::<T>::AssetIdDoesNotExist);

Something like this form.

Reply:
todo: mutate leftover, not emergency.